File: 백준/Silver/1535. 안녕/안녕.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

// 체력: 100 시작
// 체력은 0보다 커야함
// 최대 기쁨

const N = Number(input[0])
const L = input[1].split(" ").map(Number)
const J = input[2].split(" ").map(Number)

const dp = Array.from({ length: 101 }, () => 0)

for (let i = 0; i < N; i++) {
  for (let j = 100; j > 0; j--) {
    if (L[i] < j) {
      dp[j] = Math.max(dp[j - L[i]] + J[i], dp[j])
    }
  }
}

console.log(Math.max(...dp))

File: 백준/Silver/1535. 안녕/안녕.py
import sys
input = sys.stdin.readline

n = int(input())
L = list(map(int,input().split()))
J = list(map(int,input().split()))
dp  = [[0 for i in range(0,101)] for _ in range(n+1)]


for i in range(1,n+1):
  lost = L[i-1]
  joy = J[i-1]

  for hp in range(1,101):
    if(hp <=lost):
      dp[i][hp] = dp[i-1][hp]
    else:
      dp[i][hp] = max(dp[i-1][hp], dp[i-1][hp-lost]+joy)

print(max(map(max,dp)))
